spec.html

```diff
@@ -8871,18 +8871,22 @@ <h1>[[Construct]] ( _argumentsList_, _newTarget_ )</h1>
     </emu-clause>
 
     <emu-clause id="sec-createbuiltinfunction" aoid="CreateBuiltinFunction">
-      <h1>CreateBuiltinFunction ( _steps_, _internalSlotsList_ [ , _realm_ [ , _prototype_ ] ] )</h1>
-      <p>The abstract operation CreateBuiltinFunction takes arguments _steps_ and _internalSlotsList_ (a List of names of internal slots) and optional arguments _realm_ and _prototype_. _internalSlotsList_ contains the names of additional internal slots that must be defined as part of the object. This operation creates a built-in function object. It performs the following steps when called:</p>
+      <h1>CreateBuiltinFunction ( _algorithm_, _internalSlotsList_ [ , _realm_ [ , _prototype_ ] ] )</h1>
+      <p>The abstract operation CreateBuiltinFunction takes arguments _algorithm_ and _internalSlotsList_ (a List of names of internal slots) and optional arguments _realm_ and _prototype_. _internalSlotsList_ contains the names of additional internal slots that must be defined as part of the object. This operation creates a built-in function object. It performs the following steps when called:</p>
       <emu-alg>
-        1. Assert: _steps_ is either a set of algorithm steps or other definition of a function's behaviour provided in this specification.
+        1. Assert: _algorithm_ is either an Abstract Closure, a set of algorithm steps, or some other definition of a function's behaviour provided in this specification.
         1. If _realm_ is not present, set _realm_ to the current Realm Record.
         1. Assert: _realm_ is a Realm Record.
         1. If _prototype_ is not present, set _prototype_ to _realm_.[[Intrinsics]].[[%Function.prototype%]].
-        1. Let _func_ be a new built-in function object that when called performs the action described by _steps_. The new function object has internal slots whose names are the elements of _internalSlotsList_.
+        1. Let _func_ be a new built-in function object that when called performs the action described by _algorithm_. The new function object has internal slots whose names are the elements of _internalSlotsList_.
```

Comment:
I feel like this lacking clarity a bit for the case with `_algorithm_` is an Abstract Closure. Since the definition of ACs explicitly talk about invoking them, perhaps we should have a sentence here that in the case _algorithm_ is an AC, it is invoked with the same arguments in the same order that the built-in function objects was called with, and its return value returned.

Reply:
Theoretically, _algorithm_ (whether or not it's an AC) is never invoked, it merely specifies the behavior that the new function must exhibit.

Reply:
Fair enough, it is not invoked in the sense of an actual JS function invocation. There's no term we use for the "meta invocation" that's spec-internal calling of an AC. The high order bit for me here was to say something about forwarding of arguments. Perhaps nobody else finds this under-specified and considers "performs the action described" to include the argument forwarding?

Reply:
Actually, when I said "invoked", I did mean it in your sense of "meta invocation": _algorithm_ is never meta-invoked. When the new function object is called, it doesn't forward arguments to _algorithm_, because _algorithm_ isn't in the picture.

E.g., you can imagine that the implementation uses _algorithm_ as a template/blueprint for building the function, and then discards it once the function is created (i.e., by the time this step finishes). _algorithm_'s parameters are a template for the function's parameters, so when the function is called, it will handle its arguments in a way that is consistent with how _algorithm_ handles its parameters.

That's how I read it, anyway.

Reply:
That's a fine mental model, I'd still like a sentence about it, like "the function object handles parameters in a way consistent with how _algorithm_ handles parameters"

Reply:
Perhaps "performs the action described by _algorithm_ using the provided arguments as the values of the corresponding parameters specified by _algorithm_"?

That doesn't describe how to deal with the case that the function is called with the wrong number of arguments, but that's covered by these paragraphs in ECMAScript Standard Built-in Objects:

> Unless otherwise specified in the description of a particular function, if a built-in function or constructor is given fewer arguments than the function is specified to require, the function or constructor shall behave exactly as if it had been given sufficient additional arguments, each such argument being the undefined value. Such missing arguments are considered to be “not present” and may be identified in that manner by specification algorithms. In the description of a particular function, the terms “this value” and “NewTarget” have the meanings given in 10.3.
> 
> Unless otherwise specified in the description of a particular function, if a built-in function or constructor described is given more arguments than the function is specified to allow, the extra arguments are evaluated by the call and then ignored by the function. However, an implementation may define implementation specific behaviour relating to such arguments as long as the behaviour is not the throwing of a TypeError exception that is predicated simply on the presence of an extra argument.

Reply:
This wording sgtm.